Refactor keyring to split bridge logic

bergarces opened the pull request:
The goal of this PR is to split the logic that interacts directly with the Ledger device from the keyring logic that handles the interface with `KeyringController`.

- **BREAKING:** Splits keyring and bridge logic into `TrezorKeyring` and `TrezorConnectBridge` classes.
- ADDED: `TrezorBridge` interface to use in other implementations.
- CHANGED: Removes initialisation code from the constructor and to the already supported `init` method.

bergarces commented on package.json:
```diff
@@ -19,13 +19,17 @@
     "setup": "yarn install && yarn allow-scripts",
     "test": "mocha",
-    "lint:eslint": "eslint . --cache --ext js,ts",
+    "lint:eslint": "eslint . --cache --ext js,ts --ignore-pattern @types/index.d.ts",
```
I don't like the idea of ignoring it, but given that we don't have typescript support here eslint was being a bit of a hassle.

legobeat replied:
Maybe this can be added to `.eslintrc.js` instead?

bergarces commented on trezor-keyring-mv2.js:
```diff
@@ -0,0 +1,49 @@
+    // here: https://github.com/trezor/connect/blob/dec4a56af8a65a6059fb5f63fa3c6690d2c37e00/src/js/iframe/builder.js#L181
+    TrezorConnect.dispose();
+    return Promise.resolve();
+  }
```
It is worth mentioning I noticed that this method is only present for trezor and ledger keyring, but they go under different names (`dispose` and `destroy`).

Perhaps it would be a good idea to standardise the name so that we can call it if it exists. Same as we are doing with `init`.

This is where they are being called.
https://github.com/MetaMask/metamask-extension/blob/d21757c029794fb8004f2d7a36a52b1fe801c5bf/app/scripts/metamask-controller.js#L4391

Gudahtt replied:
Agreed! We've used `destroy` for this sort of thing elsewhere, so maybe that makes sense as the standard name. We can do that in a separate PR.

mcmire commented on trezor-keyring-mv2.js:
```diff
@@ -0,0 +1,49 @@
+} = require('./trezor-keyring');
+
+class TrezorKeyringMv2 extends TrezorKeyring {
+  constructor(opts = {}) {
```
Is this constructor needed? It appears to do the same thing as the superclass.

bergarces replied:
Not needed, I've removed it now.

mcmire commented on trezor-keyring-mv2.js:
```diff
@@ -0,0 +1,49 @@
+  TREZOR_CONNECT_MANIFEST,
+} = require('./trezor-keyring');
+
+class TrezorKeyringMv2 extends TrezorKeyring {
```
This class doesn't appear to be the exact same as the previous TrezorKeyring class. For instance:

* The initialization logic that was in the constructor has been moved to `init`
* `dispose` now returns a promise

Is that intentional in this PR? Those would be breaking changes if so, in which case we no longer have to have `index.js` expose a default export.

bergarces replied:
Adding the `init` method and making the `dispose` method return a promise in case that other implementations of the bridge logic require it are intentional changes. I've added a PR for the KeyringController to take the `init` method into account, as other keyring implementations might need it as well.

What I'm not too sure about is if I should make them completely optional by not mentioning them at all in the base class, return an error or return an empty promise.

I'm starting to lean towards not defining them at all in the base class, then have a separate interface in the exported types in case they have to be added so that whoever implements them has a function signature, but would like to discuss it further.

mcmire reviewed (commented):
Hey @bergarces, came across this PR in my notifications and had a few suggestions. Cheers!

bergarces commented on test/test-eth-trezor-keyring.js:
```diff
@@ -16,7 +16,8 @@ const {
 } = require('@ethereumjs/tx');
```
Changed.

bergarces commented on trezor-keyring.js:
```diff
@@ -0,0 +1,561 @@
+  return typeof tx.getChainId === 'function';
+}
+
+class TrezorKeyring extends EventEmitter {
```
It isn't meant to be used on its own, as it requires the protected methods to communicate with the device to be overriden. If we were already using TypeScript we could just define it as an abstract class, but will rename to `BaseTrezorKeyring` for clarity.

bergarces commented on @types/index.d.ts:
```diff
@@ -0,0 +1,70 @@
+export const KEYRING_TYPE: string;
```
Good point. Changed.

bergarces commented:
> Hey @bergarces, came across this PR in my notifications and had a few suggestions. Cheers!

Hey @mcmire, thanks for taking the time to review and for the suggestions, I've applied most of them and I'll make sure to use them for the ledger keyring changes too.

bergarces commented:
I'd also wanted to say... would it be better to move all source files to its own folder?

legobeat commented on base-trezor-keyring.js:
```diff
@@ -0,0 +1,562 @@
+          }
+        })
+        .catch((e) => {
+          reject(new Error((e && e.toString()) || 'Unknown error'));
```
```suggestion
          reject(new Error(e?.toString() || 'Unknown error'));
```

legobeat commented on base-trezor-keyring.js:
```diff
@@ -0,0 +1,562 @@
+  }
+
+  isUnlocked() {
+    return Boolean(this.hdk && this.hdk.publicKey);
```
```suggestion
    return Boolean(this.hdk?.publicKey);
```

legobeat commented on base-trezor-keyring.js:
```diff
@@ -0,0 +1,562 @@
+    throw new Error('Method _ethereumSignTypedData not implemented');
+  }
+
+  serialize() {
```
Nit: There is no serialization/deserialization happening so could be confusing. Maybe something like `getOptions()`/`loadOptions(opts)`, or even a unified `options(opts)`?

bergarces replied:
I don't think we can't change the names for `serialize` and `deserialize`, they are common to all the keyrings and they are being called by `KeyringController` and the extension itself in multiple places.

legobeat replied:
ACK!

legobeat commented on base-trezor-keyring.js:
```diff
@@ -0,0 +1,562 @@
+          if (response.success) {
+            this.hdk.publicKey = Buffer.from(response.payload.publicKey, 'hex');
+            this.hdk.chainCode = Buffer.from(response.payload.chainCode, 'hex');
+            resolve('just unlocked');
```
Can these strings be made into named constants so they can be referenced to in handling by calling code?

bergarces replied:
I didn't even realise how that string was being used. I'll definitely refactor that even if it's only to add the options as a constants (an enum-like in js). Too easy to make mistakes otherwise.

legobeat commented on base-trezor-keyring.js:
```diff
@@ -0,0 +1,562 @@
+    return new Promise((resolve, reject) => {
+      this.unlock()
+        .then((_) => {
+          const from = this.unlockedAccount;
+          const to = from + n;
+
+          for (let i = from; i < to; i++) {
+            const address = this._addressFromIndex(pathBase, i);
```
```suggestion
          for (let i = 0; i < n; i++) {
            const address = this._addressFromIndex(pathBase, this.unlockedAccount + i);
```
Are the `from`/`to` bookkeeping variables really helping? I find this more clear.

legobeat commented on base-trezor-keyring.js:
```diff
@@ -0,0 +1,562 @@
+
+          for (let i = from; i < to; i++) {
+            const address = this._addressFromIndex(pathBase, i);
+            if (!this.accounts.includes(address)) {
```
I'd suggest using a set rather than array and this logic becomes redundant.

legobeat commented on base-trezor-keyring.js:
```diff
@@ -0,0 +1,562 @@
+    return this.__getPage(-1);
+  }
+
+  __getPage(increment) {
+    this.page += increment;
+
+    if (this.page <= 0) {
+      this.page = 1;
+    }
+
+    return new Promise((resolve, reject) => {
+      this.unlock()
+        .then((_) => {
+          const from = (this.page - 1) * this.perPage;
+          const to = from + this.perPage;
+
+          const accounts = [];
+
+          for (let i = from; i < to; i++) {
+            const address = this._addressFromIndex(pathBase, i);
+            accounts.push({
+              address,
+              balance: null,
+              index: i,
+            });
+            this.paths[ethUtil.toChecksumAddress(address)] = i;
+          }
+          resolve(accounts);
+        })
+        .catch((e) => {
+          reject(e);
+        });
+    });
+  }
```
I think `getPage` suggests giving it a page number? Using async primitives and starting index at 0 also makes it more clear, I think.

Roughly:

```suggestion
  __paginate(increment) {
    this.page = Math.max(0, this.page + increment);
    await this.unlock();
    const accounts = [];

    for (let i = 0; i < this.perPage; i++) {
      const index = this.page * this.perPage + i;
      const address = this._addressFromIndex(pathBase, index);
      accounts.push({
        address,
        balance: null,
        index,
      });
      this.paths[ethUtil.toChecksumAddress(address)] = index;
    }

    return accounts;
  }
```

Also, is there a significance wrt `_` vs `__` as prefix for private functions?

bergarces replied:
I think we definitely should be looking into refactoring the pagination logic in a way such as the one you suggest. I was just wondering that, given how much of that code is almost identical within multiple keyrings (at least trezor and ledger), if it would be better to refactor pagination logic in a way that it can be reused.

Maybe it could be written within KeyringController and injected into individual keyrings, or maybe added to a new library along with other common tools for keyrings (e.g. hdkey for derivation).

I think we should assess if that's something we could be doing before refactoring individual keyrings. Any opinion on this @Gudahtt ?

Gudahtt replied:
Reusing the pagination logic makes a lot of sense, but I don't think it would belong at the KeyringController level because not all keyrings have pagination.

legobeat commented on base-trezor-keyring.js:
```diff
@@ -0,0 +1,562 @@
+
+  removeAccount(address) {
+    if (
+      !this.accounts.map((a) => a.toLowerCase()).includes(address.toLowerCase())
```
We don't need to iterate over all accounts every time.
```suggestion
      this.accounts.every((a) => a.toLowerCase() !== address.toLowerCase())
```

legobeat commented on base-trezor-keyring.js:
```diff
@@ -0,0 +1,562 @@
+  }
+
+  removeAccount(address) {
+    if (
+      !this.accounts.map((a) => a.toLowerCase()).includes(address.toLowerCase())
+    ) {
+      throw new Error(`Address ${address} not found in this keyring`);
+    }
+
+    this.accounts = this.accounts.filter(
+      (a) => a.toLowerCase() !== address.toLowerCase(),
+    );
+  }
```
```suggestion
    const accounts = this.accounts.filter(a =>
      t.toLowerCase() !== address.toLowerCase()
    }
    if (accounts.length === this.accounts.length) {
      throw new Error(`Address ${address} not found in this keyring`);
    }
    this.accounts = accounts;
  }
```

bergarces commented on base-trezor-keyring.js:
```diff
@@ -0,0 +1,562 @@
+          } else {
+            reject(
+              new Error(
+                (response.payload && response.payload.error) || 'Unknown error',
```
I tried using optional chaining, but the existing eslint config sets `ecmaVersion` to 2018. I believe we need 2020 for that. I'm happy to change it and use optional chaining if that doesn't cause any unwanted side effect.

legobeat replied:
Ah, https://github.com/MetaMask/eth-trezor-keyring/blob/440074fce83de77f608b759ba07eff54619abe9e/package.json#L65 so 2020 should be good?

Gudahtt replied:
This library is used in the extension, which still supports browsers that do not support optional chaining, and as a general rule we try to avoid introducing the need to transpile deps. We're working on dropping support for older browsers, and we've been migrating everything to TypeScript so that we can write using modern syntax. But for pure JS libraries it's better not to use newer syntax for now.

Gudahtt commented:
Interesting - so the "bridge" portion is the child class, extending from the base keyring with the shared logic.

Have you considered passing the bridge in as a constructor parameter? Then we could have a bridge API rather than needing the bridge to extend the keyring and override methods.

Gudahtt commented:
^ I just thought of one more advantage to passing the bridge in: it makes it feasible to change the bridge at runtime.

Not that we'd need that to be a supported feature right away, but, this relates to the idea I mentioned at one point for the Ledger keyring of splitting up the Ledger Live bride from the Ledger iframe-based bridge. That's a toggle that exists at runtime, so we'd weant the bridge to be changeable at runtime. Which... we could do if it was a child class by saving state, reconstructing, and restoring, but that would be complex to manage. It seems more sensible to make the bridge a separate entity that we could swap out without reconstructing the keyring itself.

bergarces commented on test/eth-trezor-keyring.test.js:
```diff
@@ -1,24 +1,17 @@
-global.window = require('./window.shim');
-global.navigator = require('./navigator.shim');
-global.self = require('./self.shim');
```
I'm not sure why these shims were needed in the past, but I'm pretty certain they are not needed now.

bergarces commented on trezor-keyring.js:
```diff
@@ -0,0 +1,534 @@
+}
+
+class TrezorKeyring extends EventEmitter {
+  constructor(bridge) {
```
Constructor now takes a bridge dependency instead of `opts`.

It does no longer initialize Trezor or deserialize, as those are asynchronous calls that should be called outside the constructor.

This is a breaking change and changes are needed in the KeyringController to handle them.

bergarces commented on trezor-keyring.js:
```diff
@@ -0,0 +1,534 @@
+   * @returns {"T" | "1" | undefined}
+   */
+  getModel() {
+    return this.bridge.model;
```
`model` is now a property of the bridge. It gets updated based on a trezor device event.

It could be moved back here and updated if we pass a callback to `this.bridge.init()`.

bergarces commented:
> ^ I just thought of one more advantage to passing the bridge in: it makes it feasible to change the bridge at runtime.
> 
> Not that we'd need that to be a supported feature right away, but, this relates to the idea I mentioned at one point for splitting up the Ledger Live bridge from the Ledger iframe-based bridge. That's a toggle that exists at runtime, so we'd weant the bridge to be changeable at runtime. Which... we could do if it was a child class by saving state, reconstructing, and restoring, but that would be complex to manage. It seems more sensible to make the bridge a separate entity that we could swap out without reconstructing the keyring itself.

Hey @Gudahtt, I've added the discussed changes. Instead of extending from `BaseTrezorKeyring` (now `TrezorKeyring`), the logic is now in its own `TrezorBridgeMv2` and a TypeScript interface is exported for new bridge implementations.

Since deserialize also returns a promise, even though there's nothing asynchronous within the method, I've also moved it out of the constructor and replaced the `opts` by the bridge dependency. The changes needed to handle that are in the `KeyringController` PR.

https://github.com/MetaMask/KeyringController/pull/163/files

andrewpeters9 commented on test/eth-bridge-mv2.test.js:
```diff
@@ -0,0 +1,121 @@
+const sinon = require('sinon');
```
Regarding file name, should the `mv2` terminology be dropped given that this package can (should?) be extension/manifest agnostic? (I left a similar comment on the Ledger MR)

bergarces replied:
The MV2 terminology for the bridge should be maintained, as the bridge for MV2 and MV3 are unlikely to be the same, since we cannot use `TrezorConnect` directly from within the extension.

What it's true, IMO, is that the MV2 and MV3 bridge implementations should live in the extension/mobile/desktop code and not in the keyring, as they might be using different dependencies to achieve that goal.

The only reason I left the MV2 bridge here is to make the transition easier.

andrewpeters9 replied:
@bergarces Understood, but given that there could be a non-extension-related library (i.e., a non-mv-related libary) that might want to the `mv2` bridge - i.e., the TrezorConnect bridge, then would it not be more logical to rename this keyring something like `TrezorNativeBridge` instead of `TrezorBridgeMv2`.

Bear in mind that the `TrezorConnect` library can still be used on `mv3` assuming some changes are made to `metamask-extension` i.e., persisting clients etc. My draft MR for MV3 Hardware wallets currently uses TrezorConnect, but it still gets caught on some edge-cases until this issue is resolved: https://github.com/MetaMask/metamask-extension/issues/17649

E.g., Enkrypt uses `TrezorConnect` (albeit via `@trezor/connect-web` for their `mv3` chrome extension without issues: https://github.com/enkryptcom/enKrypt/search?q=trezor%2Fconnect

andrewpeters9 commented on trezor-keyring.js:
```diff
@@ -0,0 +1,534 @@
+
+  serialize() {
+    return Promise.resolve({
+      hdPath: this.hdPath,
```
Could we add this:
```js
    hdk: this.isUnlocked() ? {
       publicKey: this.hdk.publicKey.toString('hex'),
       chainCode: this.hdk.chainCode.toString('hex'),
    } : null,
```

bergarces replied:
Hey @andrewpeters9 

I completely agree that those values should also be serialized/deserialized. But that should be done in its own PR, as I wouldn't want to add unrelated fixes to this one. The plan here is to just split the code from keyring and bridge.

I'm happy to rebase this one if that gets merged first.

andrewpeters9 replied:
@bergarces Will do, I'll open separate MRs for my changes

andrewpeters9 commented on trezor-keyring.js:
```diff
@@ -0,0 +1,534 @@
+  }
+
+  deserialize(opts = {}) {
+    this.hdPath = opts.hdPath || hdPathString;
```
Can we add this? (Idea being that all serializable data is deserialized:
```js
    if(opts.hdk){
      this.setHdk(opts.hdk);
    }

    this.paths = opts.paths || {};
    this.unlockedAccount = opts.unlockedAccount || 0;
```

and this method

```js
  setHdk ({ publicKey, chainCode }) {
    this.hdk.publicKey = Buffer.from(publicKey, 'hex')
    this.hdk.chainCode = Buffer.from(chainCode, 'hex')
  }
```

bergarces commented on src/trezor-connect-bridge.ts:
```diff
@@ -0,0 +1,64 @@
+    } & Partial<ConnectSettings>,
+  ) {
+    TrezorConnect.on(DEVICE_EVENT, (event) => {
+      if (event.type !== DEVICE.CONNECT) {
```
It's possible to force typescript type inference by using the event discriminator, that saves the use of `hasProperty` and avoid importing `@metamask/utils` as a dependency.

bergarces commented on src/trezor-keyring.test.ts:
```diff
@@ -15,16 +14,8 @@ import { Common, Chain, Hardfork } from '@ethereumjs/common';
-
-global.window = windowShim;
-global.navigator = navigatorShim;
-global.self = selfShim;
```
This shims are not needed for the tests, as we are creating stubs for `TrezorConnect`.

Gudahtt reviewed (approved):
LGTM!
